tests/UI/Component/Input/Field/PasswordInputTest.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+/* Copyright (c) 2017 Richard Klees <richard.klees@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+require_once(__DIR__ . "/../../../../../libs/composer/vendor/autoload.php");
+require_once(__DIR__ . "/../../../Base.php");
+require_once(__DIR__ . "/InputTest.php");
+
+use ILIAS\UI\Implementation\Component\SignalGenerator;
+use ILIAS\UI\Implementation\Component\Input\PostData;
+use ILIAS\Data\Password as PWD;
+
+class _PWDPostData implements PostData {
+	public function get($name) {
+		return 'some value';
+	}
+	public function getOr($name, $default) {
+		return 'some alternative value';
+	}
+
+}
+
+class PasswordInputTest extends ILIAS_UI_TestBase {
+
+	public function setUp() {
+		$this->name_source = new DefNamesource();
+	}
+
+
+	protected function buildFactory() {
+		return new ILIAS\UI\Implementation\Component\Input\Field\Factory(new SignalGenerator());
+	}
+
+
+	public function test_implements_factory_interface() {
+		$f = $this->buildFactory();
+		$pwd = $f->password("label", "byline");
+	}
+
+
+	public function test_render() {
+		$f = $this->buildFactory();
+		$label = "label";
+		$byline = "byline";
+		$name = "name_0";
+		$pwd = $f->password($label, $byline)->withNameFrom($this->name_source);
+
+		$r = $this->getDefaultRenderer();
+		$expected = ""
+			."<div class=\"form-group row\">"
+				."<label for=\"$name\" class=\"control-label col-sm-3\">$label</label>"
+				."<div class=\"col-sm-9\">"
+					."<input type=\"password\" name=\"$name\" class=\"form-control form-control-sm\" />"
+					."<div class=\"help-block\">$byline</div>"
+				."</div>"
+			."</div>";
+		$this->assertHTMLEquals($expected, $r->render($pwd));
+	}
+
+
+	public function test_render_error() {
+		$f = $this->buildFactory();
+		$label = "label";
+		$byline = "byline";
+		$name = "name_0";
+		$error = "an_error";
+		$pwd = $f->password($label, $byline)->withNameFrom($this->name_source)->withError($error);
+
+		$r = $this->getDefaultRenderer();
+		$html = $this->normalizeHTML($r->render($pwd));
+		$expected = ""
+			."<div class=\"form-group row\">"
+				." <label for=\"$name\" class=\"control-label col-sm-3\">$label</label>"
+				." <div class=\"col-sm-9\">"
+					." <input type=\"password\" name=\"$name\" class=\"form-control form-control-sm\" />"
+					." <div class=\"help-block\">$byline</div>"
+					." <div class=\"help-block alert alert-danger\" role=\"alert\">"
+						." <img border=\"0\" src=\"./templates/default/images/icon_alert.svg\" alt=\"alert\">"
+						." $error"
+					." </div>"
+				." </div>"
+			."</div>";
+
+		$html = preg_replace('!\s+!', ' ', $html);
+		$expected = preg_replace('!\s+!', ' ', $expected);
+		$html = explode( ' ', $html); //so you can actually _see_ the difference...
+		$expected = explode( ' ', $expected);
```

Comment:
Ja, looks strange...
Unfortunately, when using assertHTMLEquals, there is a weired error:
DOMDocument::loadXML(): Opening and ending tag mismatch: img line 1 and div in Entity, line: 1
that I could not get rid of; since I do not test xml-syntax here, but expected output, I opted fro strings.

Reply:
That error is due to the fact, that XML does not know the empty tags that are sometimes used in HTML.